### PR TITLE
Enable .NET analyzers

### PR DESCRIPTION
**Pull request by Saancreed**
Not sure if this is the correct resolution for all reported warnings, so I'm open to suggestions.

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -4,6 +4,10 @@
     <LangVersion>preview</LangVersion>
     <Nullable>enable</Nullable>
     <ImplicitUsings>enable</ImplicitUsings>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <AnalysisLevel>latest</AnalysisLevel>
+    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
+    <NoWarn>CA1031;CA1711</NoWarn>
```
Comment (jakubfijalkowski) [inline Directory.Build.props:10]: Can we _no_ disable these warnings globally?

File: Directory.Build.targets [hunk 2]
```diff
@@ -27,7 +25,7 @@
     <PackageReference Update="Microsoft.Extensions.DependencyModel" Version="$(ExtensionsVersion)" />
     <PackageReference Update="Microsoft.Extensions.FileSystemGlobbing" Version="$(ExtensionsVersion)" />
 
-    <PackageReference Update="NuGet.Frameworks" Version="$(NuGetVersion)" />
+    <PackageReference Update="NuGet.Frameworks" Version="*" />
```
Comment (jakubfijalkowski) [inline Directory.Build.targets:28]: Let's leave a note why this is done like it is.

File: Directory.Build.targets [hunk 3]
```diff
@@ -42,7 +40,8 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="LeanCode.CodeAnalysis" Version="$(CoreLibVersion)" PrivateAssets="all" />
+    <PackageReference Include="LeanCode.CodeAnalysis" Version="6.0.1583" PrivateAssets="all" />
+    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.435" PrivateAssets="all" />
```
Comment (jakubfijalkowski) [inline Directory.Build.targets:44]: Let's bump the StyleCop version in CoreLib _then_ let's use it here.

File: src/LeanCode.Contracts/ICommand.cs [hunk 4]
```diff
@@ -1,4 +1,4 @@
 namespace LeanCode.Contracts;
 
-[System.Diagnostics.CodeAnalysis.SuppressMessage("?", "CA1040", Justification = "Marker interface.")]
+[System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1040", Justification = "Marker interface.")]
```
Comment (jakubfijalkowski) [inline src/LeanCode.Contracts/ICommand.cs:3]: Please, let's leave `?` everywhere - the `category` introduces unnecessary noise.

File: src/LeanCode.ContractsGenerator/Analyzers/ErrorCodesUniqueness.cs [hunk 5]
```diff
@@ -4,9 +4,9 @@ public class ErrorCodesUniqueness : BaseAnalyzer
 {
     public const string Code = "CNTR0003";
 
-    public override IEnumerable<AnalyzeError> AnalyzeErrorCodes(AnalyzerContext context, IEnumerable<ErrorCode> errorCodes)
+    public override IEnumerable<AnalyzeError> AnalyzeErrorCodes(AnalyzerContext context, IEnumerable<ErrorCode> errCodes)
```
Comment (jakubfijalkowski) [inline src/LeanCode.ContractsGenerator/Analyzers/ErrorCodesUniqueness.cs:7]: Why the change?

File: src/LeanCode.ContractsGenerator/Analyzers/InternalStructureCheck.cs [hunk 6]
```diff
@@ -14,15 +14,15 @@ public override IEnumerable<AnalyzeError> AnalyzeTypeRef(AnalyzerContext context
         }
     }
 
-    public override IEnumerable<AnalyzeError> AnalyzeValueRef(AnalyzerContext context, ValueRef vr)
+    public override IEnumerable<AnalyzeError> AnalyzeValueRef(AnalyzerContext context, ValueRef valueRef)
```
Comment (jakubfijalkowski) [inline src/LeanCode.ContractsGenerator/Analyzers/InternalStructureCheck.cs:17]: Yhh...

File: src/LeanCode.ContractsGenerator/Compilation/ContractsCompiler.cs [hunk 7]
```diff
@@ -47,22 +48,18 @@ private static bool IsWantedLeanCodeAssembly(CompilationLibrary cl)
         return LeanCodeAssemblyNames.Contains(cl.Name, StringComparer.InvariantCultureIgnoreCase);
     }
 
-    public static readonly ImmutableList<PortableExecutableReference> DefaultAssemblies;
+    private static readonly Assembly ExecutingAssembly = Assembly.GetExecutingAssembly();
```
Comment (jakubfijalkowski) [inline src/LeanCode.ContractsGenerator/Compilation/ContractsCompiler.cs:51]: What was wrong with the static constructor?

File: src/LeanCode.ContractsGenerator/Program.cs [hunk 8]
```diff
@@ -66,44 +66,44 @@ private static async Task<int> Main(string[] args)
         }
         catch (InvalidProjectException ex)
         {
-            Console.Error.WriteLine($"Cannot load one of the projects: {ex.Message}");
-            Console.Error.WriteLine("At");
-            Console.Error.WriteLine(ex.StackTrace);
+            await Console.Error.WriteLineAsync($"Cannot load one of the projects: {ex.Message}");
```
Comment (jakubfijalkowski) [inline src/LeanCode.ContractsGenerator/Program.cs:69]: TBH, that's unnecessary (although correct) noise introduced by the too-generic analyzers (they don't mark `Console.WriteLine` as being wrong, don't they).

File: Directory.Build.targets (hunk 3 above)
Comment (Saancreed) [reply Directory.Build.targets:44]: It is new enough in CoreLib, but 6.0.1583 is the newest CoreLib available on NuGet. This is just me not wanting to add extra feeds just to be able to build the generator.

Comment (jakubfijalkowski) [reply Directory.Build.targets:44]: Then let's publish the version on NuGet. :P

I really don't want to have stray overrides here and there.

File: src/LeanCode.Contracts/ICommand.cs (hunk 4 above)
Comment (Saancreed) [reply src/LeanCode.Contracts/ICommand.cs:3]: Counterargument: if `category` is just unnecessary noise, then shouldn't we just switch to `#pragma warning disable` which gives us better control over which lines have warnings suppressed (while also making `Justification` optional)?

File: src/LeanCode.ContractsGenerator/Analyzers/ErrorCodesUniqueness.cs (hunk 5 above)
Comment (Saancreed) [reply src/LeanCode.ContractsGenerator/Analyzers/ErrorCodesUniqueness.cs:7]: Analyzers complain when the name used in the override is different from the one used in interface / base class.

File: src/LeanCode.Contracts/ICommand.cs (hunk 4 above)
Comment (jakubfijalkowski) [reply src/LeanCode.Contracts/ICommand.cs:3]: `pragma warning disable` requires you to have the `restore` in a correct place, which introduces even more noise and it is easy to forget about one.

Also, `Justification` is perfectly fine.

File: src/LeanCode.ContractsGenerator/Program.cs (hunk 8 above)
Comment (jakubfijalkowski) [reply src/LeanCode.ContractsGenerator/Program.cs:69]: I don't really know - I don't like this, but I also don't like additional `SuppressMessage`... :/

File: src/LeanCode.Contracts/ICommand.cs (hunk 4 above)
Comment (Saancreed) [reply src/LeanCode.Contracts/ICommand.cs:3]: Okay. I don't fully agree here but I'll go with `SuppressMessage("?"` because I don't feel like I'll be able to express by reasoning here well enough :upside_down_face:

File: Directory.Build.targets (hunk 3 above)
Comment (Saancreed) [reply Directory.Build.targets:44]: What if we _removed_ `LeanCode.CodeAnalysis` instead? This would save us weird cyclical dependencies between repositories, which, while not currently problematic, would probably cause anyone reading this for the first time to be quite confused.

Although this way we lose(?) our custom ruleset file, which is probably not optimal either…

File: src/LeanCode.ContractsGenerator/Analyzers/InternalStructureCheck.cs (hunk 6 above)
Comment (Saancreed) [reply src/LeanCode.ContractsGenerator/Analyzers/InternalStructureCheck.cs:17]: Same as above.

Well, this one I renamed manually to be more consistent with other methods.

File: Directory.Build.props (hunk 1 above)
Comment (jakubfijalkowski) [reply Directory.Build.props:10]: `Exception boundary.` is fine. ;P

File: Directory.Build.targets (hunk 3 above)
Comment (jakubfijalkowski) [reply Directory.Build.targets:44]: The main addition of LeanCode.CondeAnalysis is the shared ruleset - that is the reason we use it across all the projects.

I would prefer moving the `LeanCode.CodeAnalysis` to a separate repo than removing it.

File: Directory.Build.props (hunk 1 above)
Comment (Saancreed) [reply Directory.Build.props:10]: That solves `CA1031` but what about `CA1711`?

Comment (Saancreed) [reply Directory.Build.props:10]: I don't find a warning that my _test collection_ class has name ending with `Collection` reasonable enough to keep.

File: Directory.Build.targets (hunk 3 above)
Comment (Saancreed) [reply Directory.Build.targets:44]: :cold_sweat:

So much to do just to remove one `SuppressMessage` :slightly_frowning_face:

**Review by jakubfijalkowski — APPROVED** [review]
